### PR TITLE
Debugging WebAssembly with modern tools

### PR DESCRIPTION
**Pull request by RReverser**
What's changed, or what was fixed?
- Added a post on debugging WebAssembly.

**Target Live Date:** 2020-12-09

- [ ] This has been reviewed and approved by (NAME)
- [ ] I have run `npm test` locally and all tests pass.
- [ ] I have added the appropriate `type-something` label.
- [ ] I've staged the site and manually verified that my content displays correctly.

**CC:** @petele

### CODE REVIEW

File: src/content/en/updates/2020/12/webassembly.md [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+## The road so far {: #past }
+
+A year ago, Chrome [announced initial support](/web/updates/2019/12/webassembly)
+for native WebAssembly debugging in Chrome DevTools.
+
+We demonstrated basic stepping support and talked about opportunities
+usage of [DWARF](http://dwarfstd.org/) information instead of
+source maps opens for us in the future:
+
+    > - Resolving variable names
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md]: this seems to render unexpectedly.

File: src/content/en/updates/2020/12/webassembly.md [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+3<sup>rd</sup>-party library you have no control over, or because you’re
+running into one of those bugs that occurs only on production.
+
+To aid in those cases, we’ve made some improvements to the basic
+debugging experience, too.
+
+First of all, if you used raw WebAssembly debugging before, you might
+notice that the entire disassembly is now shown in a single file – no
+more guessing which function a **Sources** entry “wasm-53834e3e/
+wasm-53834e3e-7” possibly corresponds to.
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md:259]:
```suggestion
more guessing which function a **Sources** entry `wasm-53834e3e/
wasm-53834e3e-7` possibly corresponds to.
```

remove smart quotes & switch to code font?

File: src/content/en/updates/2020/12/webassembly.md [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+### New name generation scheme {: #names }
+
+We improved names in the disassembly view, too. Previously you’d see
+just numeric indices, or, in case of functions, no name at all.
+
+Now we’re generating names similarly to other disassembly tools, by
+using hints from the [WebAssembly name
+section](https://webassembly.github.io/spec/core/appendix/custom.html#name-section),
+import/export paths and, finally, if everything else fails, generating
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md:269]:
```suggestion
`import`/`export` paths and, finally, if everything else fails, generating
```

File: src/content/en/updates/2020/12/webassembly.md [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+them based on the type and the index of the item like $func123. You can
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md:270]:
```suggestion
them based on the type and the index of the item like `$func123`. You can
```

File: src/content/en/updates/2020/12/webassembly.md [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+want to split out this debug information into a separate WebAssembly
+file. To do that in Emscripten, pass a `-gseparate-dwarf=…` flag with
+a desired filename:
+
+```bash
+$ emcc -g -gseparate-dwarf=temp.debug.wasm temp.c -o temp.html
+```
+
+In this case, the main application will only store a filename
+“temp.debug.wasm”, and the helper extension will be able to locate and
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md:383]:
```suggestion
`temp.debug.wasm`, and the helper extension will be able to locate and
```

File: src/content/en/updates/2020/12/webassembly.md [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+When combined with optimizations like described above, this feature can
+be even used to ship almost-optimized production builds of your
+application, and later debug them with a local side file. In this case,
+we’ll additionally need to override the stored URL to help the extension
+find the side file, for example:
+
+```bash
+$ emcc -g -O3 -fno-inline -gseparate-dwarf=temp.debug.wasm -s
+SEPARATE_DWARF_URL=file://[local path to temp.debug.wasm] temp.c -o
+temp.html
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md:395]:
```suggestion
$ emcc -g -O3 -fno-inline -gseparate-dwarf=temp.debug.wasm -s \
    SEPARATE_DWARF_URL=file://[local path to temp.debug.wasm] temp.c -o \
    temp.html
```

If it's a single line command, split it with `\`

File: src/content/en/updates/2020/12/webassembly.md [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+numbers](https://en.cppreference.com/w/cpp/numeric/complex) from the
+C++ standard library.
+
+I’m going to compile it with the same `-g` flag as above to include
+debug information, and also I’ll ask Emscripten to provide the SDL2
+library and allow arbitrarily-sized memory:
+
+```bash
+$ emcc -g mandelbrot.cc -o mandelbrot.html -s USE_SDL=2 -s
+ALLOW_MEMORY_GROWTH=1
```
Comment (petele) [inline src/content/en/updates/2020/12/webassembly.md:180]:
```suggestion
$ emcc -g mandelbrot.cc -o mandelbrot.html -s USE_SDL=2 -s \
    ALLOW_MEMORY_GROWTH=1
```

**Review by petele — APPROVED** [review]

File: src/content/en/updates/2020/12/webassembly.md (hunk 1 above)
Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md]: Yup, fixed in a follow-up.

File: src/content/en/updates/2020/12/webassembly.md (hunk 3 above)
Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md:269]: Those are not code quotes, just "import/export" as generic terms.

Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md:269]: OTOH... the corresponding code instructions are named that way too. I don't know, I'm happy to merge if you think this is better, but in the context it was meant as just a generic term.

Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md:270]: I think I'll leave this one as-is.

File: src/content/en/updates/2020/12/webassembly.md [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+memory.
+
+### Memory inspection {: #memory-inspector }
+
+Previously, you could only expand the WebAssembly memory object –
+represented by `env.memory` in the **Scope** view – to look up
+individual bytes. This worked in some trivial scenarios, but wasn’t
+particularly convenient to expand and didn’t allow to reinterpret data
+in formats other than byte values. We’ve added a new feature to help
+with this, too: a linear memory inspector.
```
Comment (chybie) [inline src/content/en/updates/2020/12/webassembly.md:287]: Bold it

Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md:287]: It's not a name of the section / pane though, just a description of the feature. The pane is called "Memory Inspector", I could use that, but then "linear" would be out of place.

File: src/content/en/updates/2020/12/webassembly.md [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+- Adding support for more languages.
+- …and more!
+
+Meanwhile, please help us out by trying the current beta on your own code and reporting any found
+issues to
+[https://bugs.chromium.org/p/chromium/issues/entry?template=DevTools+issue](https://bugs.chromium.org/p/chromium/issues/entry?template=DevTools+issue).
+
+Stay tuned for future updates!
+
+## Feedback {: #feedback .hide-from-toc }
```
Comment (chybie) [inline src/content/en/updates/2020/12/webassembly.md:430]: DevTools blog use different footer. Refer to here:
https://github.com/google/WebFundamentals/pull/9163/files#diff-cbc2614d28097dd7a29a74aee0e3e6c91f4efe13efbb4fd9db6902f08dd3e3d5R26

Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md:430]: So... this is a DevTools blog item then? I guess I'm confused, I thought we want to use same generic format as update last year (https://developers.google.com/web/updates/2019/12/webassembly).

**Comment by RReverser** [comment]
> LGTM, with two comments:
> 
> * Optimize the images with imageoptim, tinypng.com or squoosh.
> * Add an entry in updates/shared/discover-devtools-blog.md

Done and done.

File: src/content/en/updates/2020/12/webassembly.md (hunk 8 above)
Comment (chybie) [reply src/content/en/updates/2020/12/webassembly.md:287]: Oh, I always thought "Linear memory inspector" is the panel name. Erm.. I think we should mention the panel name "Memory inspector" somewhere in this article. Probably rephrase the sentence here or mention that in the step?

File: src/content/en/updates/2020/12/webassembly.md (hunk 9 above)
Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md]: Added.

File: src/content/en/updates/2020/12/webassembly.md (hunk 8 above)
Comment (RReverser) [reply src/content/en/updates/2020/12/webassembly.md:287]: Mentioned in the step below, let me know what you think.

**Review by chybie — APPROVED** [review]

**Pull request merged**